src/fields/field_value_cache.hh

```diff
@@ -0,0 +1,150 @@
+/*!
+ *
+﻿ * Copyright (C) 2015 Technical University of Liberec.  All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License version 3 as published by the
+ * Free Software Foundation. (http://www.gnu.org/licenses/gpl-3.0.en.html)
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+ *
+ *
+ * @file    field_value_cache.hh
+ * @brief
+ * @author  David Flanderka
+ */
+
+#ifndef FIELD_VALUE_CACHE_HH_
+#define FIELD_VALUE_CACHE_HH_
+
+#include <set>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+#include "system/armor.hh"
+
+class EvalPoints;
+class EvalSubset;
+class DHCellAccessor;
+
+
+template<class Value>
+class FieldValueCache {
+public:
+	typedef typename Value::element_type elm_type;
+
+    /// Default constructor
+    FieldValueCache();
+
+    /// Constructor
+    FieldValueCache(const EvalPoints *eval_points, unsigned int n_cache_points);
+
+    /// Destructor
+    ~FieldValueCache();
+
+    /// Marks the used local points
+    void mark_used(EvalSubset sub_set);
+
+    /// Getter for used_points
+    inline const std::set<int> &used_points() const {
+        return used_points_;
+    }
+
+    /// Return size of data cache (number of stored field values)
+    inline unsigned int size() const {
+        return data_.n_vals();
+
+    }
+
+    /// Return data vector.
+    inline Armor::Array<elm_type> &data() {
+        return data_;
+    }
+
+    /// Return data vector.
+    inline const EvalPoints *eval_points() const {
+        return eval_points_;
+    }
+
+private:
+    /// Data cache
+    Armor::Array<elm_type> data_;
+
+    /// Holds indices of used local points
+    std::set<int> used_points_;  // TODO: test unorderd_set during tuning of the performance
```

Comment:
!! Global rename `block` -> `subset`
1. rename to used_subsets
2. use fixed size array
3. declare  `MAX_N_SUBSETS` and check it in EvalPoints.
4. All side points of single eval subset is a single block in this context.

Reply:
Points 2. and 3. not solved yet.